install/tk.sh

```diff
@@ -1,50 +1,83 @@
 #!/usr/bin/env bash
-
 set -e
 
 DEFAULT_BIN_DIR="/usr/local/bin"
 BIN_DIR=${1:-"$DEFAULT_BIN_DIR"}
 
-opsys=""
-if [[ "$OSTYPE" == linux* ]]; then
-  opsys=linux
-elif [[ "$OSTYPE" == darwin* ]]; then
-  opsys=darwin
+# Determine OS and CPU architecture
+if [[ $(uname) = Darwin ]]; then
+    OS=darwin
+elif [[ $(uname) = Linux ]]; then
+    OS=linux
+else
+    OS=unknown
 fi
 
-if [[ "$opsys" == "" ]]; then
-  echo "OS $OSTYPE not supported"
-  exit 1
+case $(uname -m) in
+x86_64)
+    ARCH=amd64
+    ;;
+x86|i386|i686)
+    ARCH=386
+    ;;
+arm|armv7*)
+    ARCH=armv6
+    ;;
+arm64|aarch64*|armv8*)
+    ARCH=arm64
+    ;;
+*)
+    ARCH=unknown
+    ;;
+esac
+
+if [[ $OS = unknown ]] || [[ $ARCH = unknown ]]; then
+    echo "Unable to detect your OS and CPU architecture."
+    exit 1
+fi
+if [[ $ARCH != amd64 ]]; then
+    echo "CPU architecture $ARCH is not supported."
+    exit 1
 fi
 
 if [[ ! -x "$(command -v curl)" ]]; then
-    echo "curl not found"
+    echo "Unable to find curl"
     exit 1
 fi
 
-tmpDir=`mktemp -d`
-if [[ ! "$tmpDir" || ! -d "$tmpDir" ]]; then
-  echo "could not create temp dir"
+# Create temporary directory and trap exit to cleanup
+TMP_DIR=$(mktemp -d)
+if [[ ! $TMP_DIR || ! -d $TMP_DIR ]]; then
+  echo "Unable to create a temporary directory."
   exit 1
 fi
 
 function cleanup {
-  rm -rf "$tmpDir"
+  rm -rf "${TMP_DIR}"
 }
-
 trap cleanup EXIT
 
-pushd $tmpDir >& /dev/null
+pushd "${TMP_DIR}" >& /dev/null
+
+# Fetch repository metadata
+METADATA=$(curl -sL "https://api.github.com/repos/fluxcd/toolkit/releases/latest")
+
+# Download binary file that matches OS and CPU architecture
+DOWNLOAD_URLS=$(echo "${METADATA}" | grep browser_download_url | cut -d '"' -f 4)
+BINARY_URL=$(echo "${DOWNLOAD_URLS}" | grep "${OS}" | grep "${ARCH}")
+curl -sL -o tk.tar.gz "${BINARY_URL}"
 
-curl -s https://api.github.com/repos/fluxcd/toolkit/releases/latest |\
-  grep browser_download |\
-  grep $opsys |\
-  cut -d '"' -f 4 |\
-  xargs curl -sL -o tk.tar.gz
+# Verify that checksum matches
+CHECKSUMS_URL=$(echo "${DOWNLOAD_URLS}" | grep checksums)
+EXPECTED_CHECKSUM=$(curl -sL "${CHECKSUMS_URL}" | grep "${OS}" | grep "${ARCH}" | cut -d ' ' -f 1)
+if [[ $EXPECTED_CHECKSUM != "$(sha256sum tk.tar.gz | awk '{print $1}')" ]]; then
```

Comment:
We have to check if `sha256sum` is installed, if not we should switch to `shasum -a 256` or error out if `shasum` is not installed.

Reply:
I rewrote the script and added support for shasum fallback.